File: android/src/main/java/app/kaeru/data/report/Reporting.kt
package app.kaeru.data.report

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import com.google.firebase.FirebaseApp
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics

/**
 * What the app tells its developer about itself: which screens are opened, when an episode starts
 * or fails, how a shared viewing goes — and every crash, with the stack that caused it.
 *
 * Firebase underneath, and only when the build has a project to report to: a checkout without
 * `google-services.json` builds the same app, and every call here is then nothing at all.
 *
 * **What never leaves the phone.** No room id, no room key, no nickname, no chat line, no token,
 * nothing typed. What an event carries is a catalogue id, an episode number, a reason from a fixed
 * list — facts about the app, not about the person holding it.
 *
 * **Off means off from the first frame.** The manifest starts both collectors disabled, and they
 * are switched on here, in `Application.onCreate`, only when the viewer has not said no. Deciding
 * it the other way round — on by default, off once the setting was read — would have sent the
 * first seconds of every launch regardless.
 *
 * An object rather than something injected, for the same reason as `TogetherLog`: the player, the
 * shared session and the navigation all want to say one line from wherever they are, and none of
 * them should need a constructor parameter for it. Unit tests never install it, so it stays silent
 * there.
 */
object Reporting {
    private const val TAG = "Reporting"
    private const val PREFERENCES = "kaeru.reporting"
    private const val KEY = "enabled"

    /** Event names: lowercase and underscores, as Firebase wants them, and never more than these. */
    const val PLAY_START = "play_start"
    const val PLAY_ERROR = "play_error"
    const val TOGETHER = "together"

    private var preferences: SharedPreferences? = null
    private var analytics: FirebaseAnalytics? = null
    private var crashlytics: FirebaseCrashlytics? = null

    /** On unless the viewer turned it off. Off where nothing was installed — in tests. */
    val enabled: Boolean get() = preferences?.getBoolean(KEY, true) ?: false

    /** Whether this build can report at all: it was built with a Firebase project. */
    val available: Boolean get() = analytics != null

    fun install(context: Context, television: Boolean) {
        preferences = context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
        // FirebaseInitProvider has run by now when the build carries a config, and not otherwise.
        if (FirebaseApp.getApps(context).isEmpty()) return
        runCatching {
            analytics = FirebaseAnalytics.getInstance(context)
            crashlytics = FirebaseCrashlytics.getInstance()
        }.onFailure { Log.w(TAG, "Firebase is linked but would not start", it) }
        apply(enabled)
        // One phone and one television are two different apps to use, and the numbers are no use
        // mixed.
        analytics?.setUserProperty("device", if (television) "tv" else "phone")
    }

    fun setEnabled(on: Boolean) {
        preferences?.edit()?.putBoolean(KEY, on)?.apply()
        apply(on)
    }

    private fun apply(on: Boolean) {
        analytics?.setAnalyticsCollectionEnabled(on)
        crashlytics?.isCrashlyticsCollectionEnabled = on
    }

    /**
     * A screen came up. [name] is a route template or a fixed name — `details/{animeId}`, never
     * the id filled in — so the list of screens stays a list of screens.
     */
    fun screen(name: String) {
        if (!enabled) return
        analytics?.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, Bundle().apply {
            putString(FirebaseAnalytics.Param.SCREEN_NAME, name)
            putString(FirebaseAnalytics.Param.SCREEN_CLASS, name)
        })
        crashlytics?.log("screen $name")
    }

    /** One thing that happened, with at most a handful of plain values. Nulls are left out. */
    fun event(name: String, vararg params: Pair<String, Any?>) {
        if (!enabled) return
        val bundle = Bundle()
        params.forEach { (key, value) ->
            when (value) {
                null -> Unit
                is Int -> bundle.putLong(key, value.toLong())
                is Long -> bundle.putLong(key, value)
                is Boolean -> bundle.putString(key, value.toString())
                else -> bundle.putString(key, value.toString().take(100))
            }
        }
        analytics?.logEvent(name, bundle)
        crashlytics?.log("$name ${params.filter { it.second != null }.joinToString { "${it.first}=${it.second}" }}")
    }

    /**
     * Something that went wrong without taking the app down, and is worth a stack: a stream that
     * would not resolve, a player that gave up. Shows up in Crashlytics beside the crashes.
     */
    fun problem(error: Throwable) {
        if (!enabled) return
        crashlytics?.recordException(error)
    }
}

File: ios/Core/Reporting.swift
import Foundation
#if KAERU_FIREBASE
import FirebaseAnalytics
import FirebaseCore
import FirebaseCrashlytics
#endif

/// What the app tells its developer about itself: which screens are opened, when an episode starts
/// or fails, how a shared viewing goes — and every crash, with the stack that caused it.
///
/// Firebase underneath, and only in a build that has a project to report to: without
/// `GoogleService-Info.plist` the generator leaves `KAERU_FIREBASE` unset and every call here is
/// nothing at all. The same events, with the same names and parameters, as Android's
/// `data.report.Reporting` — one project, one set of numbers.
///
/// **What never leaves the phone.** No room id, no room key, no nickname, no chat line, no token,
/// nothing typed. An event carries a catalogue id, an episode number, a reason from a fixed list.
///
/// **Off means off from the first frame.** Info.plist starts both collectors disabled, and
/// `install()` switches them on only when the viewer has not said no in Settings.
///
/// No stored state of its own — the preference is in `UserDefaults` and Firebase keeps its own —
/// so it is callable from anywhere without an actor to hop to. Tests never install it.
enum Reporting {
    static let playStart = "play_start"
    static let playError = "play_error"
    static let together = "together"

    static let key = "reporting"

    /// On unless the viewer turned it off.
    static var enabled: Bool { UserDefaults.standard.object(forKey: key) as? Bool ?? true }

    /// Whether this build can report at all: it was built with a Firebase project, and started it.
    static var available: Bool {
        #if KAERU_FIREBASE
        return FirebaseApp.app() != nil
        #else
        return false
        #endif
    }

    static func install() {
        #if KAERU_FIREBASE
        guard Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist") != nil else { return }
        if FirebaseApp.app() == nil { FirebaseApp.configure() }
        apply(enabled)
        // The same property Android sets: an iPad and a phone are one kind of use, a television
        // another, a Mac a third, and the numbers are no use mixed.
        #if os(iOS)
        Analytics.setUserProperty("ios", forName: "device")
        #else
        Analytics.setUserProperty("macos", forName: "device")
        #endif
        #endif
    }

    static func setEnabled(_ on: Bool) {
        UserDefaults.standard.set(on, forKey: key)
        apply(on)
    }

    private static func apply(_ on: Bool) {
        #if KAERU_FIREBASE
        guard FirebaseApp.app() != nil else { return }
        Analytics.setAnalyticsCollectionEnabled(on)
        Crashlytics.crashlytics().setCrashlyticsCollectionEnabled(on)
        #endif
    }

    /// A screen came up. A fixed name — `details`, never the title's id — so the list of screens
    /// stays a list of screens.
    static func screen(_ name: String) {
        #if KAERU_FIREBASE
        guard enabled, FirebaseApp.app() != nil else { return }
        Analytics.logEvent(AnalyticsEventScreenView, parameters: [
            AnalyticsParameterScreenName: name,
            AnalyticsParameterScreenClass: name,
        ])
        Crashlytics.crashlytics().log("screen \(name)")
        #endif
    }

    /// One thing that happened, with at most a handful of plain values. Nils are left out.
    static func event(_ name: String, _ parameters: [String: Any?] = [:]) {
        #if KAERU_FIREBASE
        guard enabled, FirebaseApp.app() != nil else { return }
        var values: [String: Any] = [:]
        for (key, value) in parameters {
            switch value {
            case nil: continue
            case let number as Int: values[key] = number
            case let number as Int64: values[key] = number
            case let flag as Bool: values[key] = flag ? "true" : "false"
            case let some?: values[key] = String(String(describing: some).prefix(100))
            }
        }
        Analytics.logEvent(name, parameters: values)
        Crashlytics.crashlytics().log("\(name) \(values.map { "\($0.key)=\($0.value)" }.sorted().joined(separator: ", "))")
        #endif
    }

    /// Something that went wrong without taking the app down, and is worth a record beside the
    /// crashes: a stream that would not resolve for a reason nobody expected.
    static func problem(_ error: Error) {
        #if KAERU_FIREBASE
        guard enabled, FirebaseApp.app() != nil else { return }
        Crashlytics.crashlytics().record(error: error)
        #endif
    }
}
